risc0/circuit/rv32im-m3/src/prove.rs

```diff
@@ -24,48 +27,24 @@ use crate::verify::verify;
 
 pub type Seal = Vec<u32>;
 
-pub trait SegmentProver {
-    fn prove(&self, segment: &Segment) -> Result<Seal>;
+pub trait SegmentProver: Send + Sync {
+    fn load_segment(&self, segment: &Segment) -> Result<()>;
+
+    fn preflight(&self) -> Result<bool>;
+
+    fn prove(&self) -> Result<Seal>;
 }
 
 struct SegmentProverImpl {
     ctx: *const ProverContext,
 }
 
-impl SegmentProver for SegmentProverImpl {
-    fn prove(&self, segment: &Segment) -> Result<Seal> {
-        tracing::debug!("{segment:#?}");
-
-        // segment.partial_image.dump();
-
-        self.load_segment(segment)?;
-        self.preflight()?;
-        self.prove()?;
-
-        let transcript = self.transcript()?;
-        verify(&transcript)?;
-
-        Ok(transcript)
-    }
-}
-
-impl SegmentProverImpl {
-    #[allow(dead_code)]
-    fn new_cpu(po2: usize) -> Self {
-        Self {
-            ctx: unsafe { risc0_circuit_rv32im_m3_prover_new_cpu(po2) },
-        }
-    }
-
-    #[cfg(feature = "cuda")]
-    fn new_cuda(po2: usize) -> Self {
-        Self {
-            ctx: unsafe { risc0_circuit_rv32im_m3_prover_new_cuda(po2) },
-        }
-    }
+unsafe impl Send for SegmentProverImpl {}
+unsafe impl Sync for SegmentProverImpl {}
```

Comment:
is this correct? It would mean I think that either nothing is mutated in the `ProverContext` or else it has internal locking.